File: Day 10/solution.py
def part_one(adapters):
    effective = 0
    t, o = (0, 0)
    
    while adapters:
        if effective + 1 in adapters:
            effective += 1
            o += 1
        elif effective + 3 in adapters:
            effective += 3
            t += 1
        adapters.remove(effective)

    return t * o


def part_two(adapters):
    adapters.append(0)
    adapters.sort()

    paths = [0] * (max(adapters) + 1)
    paths[0] = 1

    for i in range(1, max(adapters) + 1):
        for j in range(1, 4):
            if i - j in adapters:
                paths[i] += paths[i - j]

    return paths[-1]
    

if __name__ == '__main__':
    inp = [int(i) for i in open("input.txt").read().split("\n")]
    inp.append(max(inp)+3)
    
    print("Part 1:", part_one(inp.copy()))
    print("Part 2:", part_two(inp.copy()))

File: Day 10/solution.rs

fn main() {
    let inp = load_input("input.txt");
    println!("Part 1: {}", part_one(&inp));
}

fn load_input(filename: &str) -> Vec<i32> {
    let file = std::fs::read_to_string(filename)
        .expect("Couldn't open the file");

    file
        .lines()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn part_one(adapters: &Vec<i32>) -> i32 {
    let mut adapters = adapters.to_vec();
    adapters.push(adapters.iter().max().unwrap() + 3);

    let mut effective = 0;
    let mut tdiff = 0;
    let mut odiff = 0;
    while !adapters.is_empty() {
        if adapters.contains(&(effective + 1)) {
            effective += 1;
            odiff += 1;
        } else if adapters.contains(&(effective + 3)) {
            effective += 3;
            tdiff += 1;
        }
        adapters.remove(adapters.iter().position(|x| x == &effective).unwrap());
    }
    tdiff * odiff
}